seccomp_test.go

```diff
@@ -701,3 +677,189 @@ func TestCreateActKillProcessFilter(t *testing.T) {
 		t.Errorf("Filter created by NewFilter was not valid")
 	}
 }
+
+//
+// Seccomp notification tests
+//
+
+// notifTest describes a seccomp notification test
+type notifTest struct {
+	syscall    ScmpSyscall
+	args       []string
+	arch       ScmpArch
+	respErr    int32
+	respVal    uint64
+	respFlags  uint32
+	syscallRet error
+}
+
+// charPtrToStr retrives the string pointer to by arg
+func charPtrToStr(ptr uint64) string {
+	var buf bytes.Buffer
+	var uptr unsafe.Pointer
+
+	for i := 0; i < 1024; i++ {
+		uptr = unsafe.Pointer(uintptr(ptr + uint64(i)))
+		b := *((*byte)(uptr))
+		if b == '\x00' {
+			break
+		}
+		buf.WriteByte(b)
+	}
+	return buf.String()
+}
+
+// notifHandler handles seccomp notifications and responses
+func notifHandler(ch chan error, fd ScmpFd, tests []notifTest) {
+
+	for _, test := range tests {
+
+		req, err := NotifReceive(fd)
+		if err != nil {
+			ch <- fmt.Errorf("Error in NotifReceive(): %s", err)
+			return
+		}
+
+		if req.Data.Syscall != test.syscall {
+			want, _ := test.syscall.GetName()
+			got, _ := req.Data.Syscall.GetName()
+			ch <- fmt.Errorf("Error in notification request syscall: got %s, want %s", got, want)
+			return
+		}
+
+		if req.Data.Arch != test.arch {
+			ch <- fmt.Errorf("Error in notification request arch: got %s, want %s", req.Data.Arch, test.arch)
+			return
+		}
+
+		for i, arg := range test.args {
+			reqArg := charPtrToStr(req.Data.Args[i])
+			if arg != reqArg {
+				ch <- fmt.Errorf("Error in syscall arg[%d]: got %s, want %s", i, reqArg, arg)
+				return
+			}
+		}
+
+		// TOCTOU check
+		if err := NotifIdValid(fd, req.Id); err != nil {
+			ch <- fmt.Errorf("TOCTOU check failed: req.Id is no longer valid: %s", err)
+			return
+		}
+
+		resp := &ScmpNotifResp{
+			Id:    req.Id,
+			Error: test.respErr,
+			Val:   test.respVal,
+			Flags: test.respFlags,
+		}
+
+		if err = NotifRespond(fd, resp); err != nil {
+			ch <- fmt.Errorf("Error in notification response: %s", err)
+			return
+		}
+
+		ch <- nil
+	}
+}
+
+func TestNotif(t *testing.T) {
+
+	// seccomp notification requires API level >= 5
+	api, err := GetApi()
+	if err != nil {
+		t.Errorf("Error getting API level: %s", err)
+	} else if api < 5 {
+		err = SetApi(5)
+		if err != nil {
+			t.Errorf("Error setting API level to 5: %s", err)
+			return
+		}
+	}
+
+	arch, err := GetNativeArch()
+	if err != nil {
+		t.Errorf("Error in GetNativeArch(): %s", err)
+		return
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Errorf("Error in Getwd(): %s", err)
+		return
+	}
+
+	filter, err := NewFilter(ActAllow)
+	if err != nil {
+		t.Errorf("Error creating filter: %s", err)
+	}
+	defer filter.Release()
+
+	// Seccomp notification is only supported on single-thread filters
+	err = filter.SetTsync(false)
```

Comment:
Linux v5.7 adds support for TSYNC and USER_NOTIF together:
https://github.com/torvalds/linux/commit/51891498f2da78ee64dfad88fa53c9e85fb50abf

It would be nice to test that.